leader/undirected_graph/sync_casteigts_metivier_robson_zemmari/sync_casteigts_metivier_robson_zemmari.go

```diff
@@ -0,0 +1,551 @@
+package sync_casteigts_metivier_robson_zemmari
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"math/bits"
+	"math/rand"
+	"time"
+
+	"github.com/krzysztof-turowski/distributed-framework/lib"
+)
+
+// Run the Casteigts-Métivier-Robson-Zemmari deterministic leader election algorithm on a random network
+func Run(n int, p float64) (int, int) {
+	nodes, synchronizer := buildSynchronizedRandomConnectedGraph(n, p)
+	rng := rand.New(rand.NewSource(time.Now().UnixNano()))
+	identifiers := rng.Perm(n)
+	for i, node := range nodes {
+		log.Println("Node", node.GetIndex(), "about to run")
+		go run(node, 2*(identifiers[i]+1)) // Every node is assigned a unique identifier of O(log(n)) bits
+	}
+	synchronizer.Synchronize(0)
+	check(nodes)
+	return synchronizer.GetStats()
+}
+
+func buildSynchronizedRandomConnectedGraph(n int, p float64) ([]lib.Node, lib.Synchronizer) {
+	possibleEdges := (n * (n - 1)) / 2
+	m := int(float64(possibleEdges) * p)
+	m = max(m, n-1)
+	m = min(m, possibleEdges)
+	weightedNodes, synchronizer := lib.BuildSynchronizedRandomConnectedWeightedGraph(n, m, 1, lib.GetGenerator())
+	nodes := make([]lib.Node, len(weightedNodes))
+	for i, weightedNode := range weightedNodes {
+		nodes[i] = weightedNode
+	}
+	return nodes, synchronizer
+}
+
+/* CHECKS */
+
+func check(nodes []lib.Node) {
+	maxIndex := 0
+	leader := none
+	for _, node := range nodes {
+		s := getState(node)
+		maxIndex = max(maxIndex, s.PlainID)
+		if !s.Elected {
+			continue
+		}
+		if leader != none {
+			panic("Multiple leaders")
+		}
+		leader = s.PlainID
+	}
+	if leader == none {
+		panic("No leader")
+	}
+	if leader != maxIndex {
+		panic("The leader does not have the greatest index")
+	}
+	for _, node := range nodes {
+		if !equal(getState(node).Prefix, toAlphaEncoding(maxIndex)) {
+			panic("A node does not know the leader's identifier")
+		}
+	}
+}
+
+/* THE SKELETON OF THE ALGORITHM */
+
+func run(node lib.Node, id int) {
+	node.StartProcessing()
+	s := newState(id, node.GetOutChannelsCount())
+	setState(node, s)
+	node.FinishProcessing(false)
+
+	finish := false
+	for round := 1; !finish; round++ {
+		node.StartProcessing()
+		finish = process(node, round)
+		node.FinishProcessing(finish)
+	}
+}
+
+func process(node lib.Node, round int) bool {
+	s := getState(node)
+
+	operation := s.updatePrefixAndParent(round)
+	announceModification(node, operation, &s)
+	if operation == shutdown {
+		node.IgnoreFutureMessages()
+		return true
+	}
+	processIncomingModifications(node, &s)
+
+	announceTermination(node, &s)
+	processIncomingTerminations(node, &s)
+
+	candidate := s.isCandidate()
+
+	if !s.Done && s.shouldSetTermination() {
+		s.Termination = true
+	}
+	reportTerminationToParent(node, &s)
+	collectTerminationsFromChildren(node, &s, candidate)
+
+	setState(node, s)
+
+	return false
+}
```

Comment:
Staying in the spirit of the repository, I'd introduce another enum ("Phase"?) and process one round (send-receive) at a time.